RecoTracker/LSTCore/src/alpaka/Segment.h

```diff
@@ -12,168 +14,65 @@
-          score_buf(allocBufWrapper<float>(devAccIn, maxPixelSegments, queue)),
-          circleCenterX_buf(allocBufWrapper<float>(devAccIn, maxPixelSegments, queue)),
-          circleCenterY_buf(allocBufWrapper<float>(devAccIn, maxPixelSegments, queue)),
-          circleRadius_buf(allocBufWrapper<float>(devAccIn, maxPixelSegments, queue)) {
-      alpaka::memset(queue, nSegments_buf, 0u);
-      alpaka::memset(queue, totOccupancySegments_buf, 0u);
-      alpaka::memset(queue, partOfPT5_buf, false);
-      alpaka::memset(queue, pLSHitsIdxs_buf, 0u);
-    }
+  GENERATE_SOA_LAYOUT(SegmentsMemSoALayout,
+    SOA_COLUMN(FPX, dPhis),
+    SOA_COLUMN(FPX, dPhiMins),
+    SOA_COLUMN(FPX, dPhiMaxs),
+    SOA_COLUMN(FPX, dPhiChanges),
+    SOA_COLUMN(FPX, dPhiChangeMins),
+    SOA_COLUMN(FPX, dPhiChangeMaxs),
+    SOA_COLUMN(uint16_t, innerLowerModuleIndices),
+    SOA_COLUMN(uint16_t, outerLowerModuleIndices),
+    SOA_COLUMN(Params_LS::ArrayUxLayers, mdIndices),
+    SOA_COLUMN(unsigned int, innerMiniDoubletAnchorHitIndices),
+    SOA_COLUMN(unsigned int, outerMiniDoubletAnchorHitIndices),
+    SOA_SCALAR(unsigned int, nMemoryLocations))
+
+  GENERATE_SOA_LAYOUT(SegmentsModSoALayout,
+    SOA_COLUMN(unsigned int, nSegments), //number of segments per inner lower module
+    SOA_COLUMN(unsigned int, totOccupancySegments))
+
+  GENERATE_SOA_LAYOUT(SegmentsPixSoALayout,
+    SOA_COLUMN(unsigned int, seedIdx),
+    SOA_COLUMN(int, charge),
+    SOA_COLUMN(int, superbin),
+    SOA_COLUMN(uint4, pLSHitsIdxs),
+    SOA_COLUMN(PixelType, pixelType),
+    SOA_COLUMN(char, isQuad),
+    SOA_COLUMN(char, isDup),
+    SOA_COLUMN(bool, partOfPT5),
+    SOA_COLUMN(float, ptIn),
+    SOA_COLUMN(float, ptErr),
+    SOA_COLUMN(float, px),
+    SOA_COLUMN(float, py),
+    SOA_COLUMN(float, pz),
+    SOA_COLUMN(float, etaErr),
+    SOA_COLUMN(float, eta),
+    SOA_COLUMN(float, phi),
+    SOA_COLUMN(float, score),
+    SOA_COLUMN(float, circleCenterX),
+    SOA_COLUMN(float, circleCenterY),
+    SOA_COLUMN(float, circleRadius))
+
+  using SegmentsMemSoA = SegmentsMemSoALayout<>;
+  using SegmentsModSoA = SegmentsModSoALayout<>;
+  using SegmentsPixSoA = SegmentsPixSoALayout<>;
+
+  using SegmentsDeviceCollection = PortableCollection3<SegmentsMemSoA, SegmentsModSoA, SegmentsPixSoA>;
+  using SegmentsHostCollection = PortableHostMultiCollection<SegmentsMemSoA, SegmentsModSoA, SegmentsPixSoA>;
 
-    inline Segments const* data() const { return &data_; }
-    inline void setData(SegmentsBuffer& buf) { data_.setData(buf); }
+  struct Segments {
+    SegmentsMemSoA::View mem;
+    SegmentsModSoA::View mod;
+    SegmentsPixSoA::View pix;
+
+    template <typename TDev>
+    void setData(PortableMultiCollection<TDev, SegmentsMemSoA, SegmentsModSoA, SegmentsPixSoA>& segmentsCol) {
+      mem = segmentsCol.template view<SegmentsMemSoA>();
+      mod = segmentsCol.template view<SegmentsModSoA>();
+      pix = segmentsCol.template view<SegmentsPixSoA>();
+    }
   };
```

Comment:
@slava77 Should I get rid of this struct and only pass around View/ConstView objects? I was doing that first, but I thought it might be convenient to keep them together and imitate the segments+segmentsBuffers structure that we had. There's definitely downsides like not making clear what should be const, so maybe we should stick with views.

Reply:
I can't recall our use cases offhand. IIRC, all 3 are not used together all the time; ... are the occupancies even used at all after the ranges are computed. After that pLS and LS are not always appearing together. So, a split may be more clear.